File: backend/text_handler/image.go
package text_handler

import (
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"time"

	"github.com/gin-gonic/gin"

	"novel2video/backend"
	"novel2video/backend/image"
	"novel2video/backend/util"
)

func GenerateImage(c *gin.Context) {
	err := os.RemoveAll(util.ImageDir)
	if err != nil {
		backend.HandleError(c, http.StatusInternalServerError, "Failed to remove directory", err)
		return
	}
	err = os.MkdirAll(util.ImageDir, os.ModePerm)
	if err != nil {
		backend.HandleError(c, http.StatusInternalServerError, "Failed to create directory", err)
		return
	}
	lines, err := readLinesFromDirectory("temp/prompts")
	if err != nil {
		backend.HandleError(c, http.StatusInternalServerError, "Failed to read fragments", err)
		return
	}
	go func() {
		for i, p := range lines {
			pi := p
			err := image.GenerateImage(pi, 114514191981, 540, 960, i)
			if err != nil {
				fmt.Println("Error:", err)
			}
		}
	}()
}

func GetLocalImages(c *gin.Context) {
	files, err := os.ReadDir(util.ImageDir)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "读取图像目录失败"})
		return
	}
	imageMap := make(map[string]string)
	re := regexp.MustCompile(`(\d+)\.png`) // 从文件名中提取数字
	now := time.Now().Unix()
	for _, file := range files {
		if !file.IsDir() {
			matches := re.FindStringSubmatch(file.Name())
			if len(matches) > 1 {
				key := matches[1]
				absPath := filepath.Join("/images", file.Name())
				imageMap[key] = absPath + fmt.Sprintf("?v=%d", now)
			}
		}
	}

	c.JSON(http.StatusOK, imageMap)
}

File: backend/rest_handler/image.py
import asyncio
from flask import Flask, jsonify, request
import os
import re
import time

from backend.image.sd import generate_image
from backend.util.constant import image_dir, prompts_en_dir
from backend.util.file import make_dir, read_lines_from_directory, remove_all

def handle_error(message, err):
    return jsonify({"error": message}), 500

async def async_generate_images(lines):
    for i, p in enumerate(lines):
        try:
            await generate_image(p, 114514191981, 540, 960, i)
        except Exception as e:
            print("Error:", e)

async def async_generate_image_single(content, index):
    try:
        await generate_image(content, 114514191981, 540, 960, index)
    except Exception as e:
        print("Error:", e)
                
def generate_images():
    try:
        remove_all(image_dir)
        make_dir(image_dir)
    except Exception as e:
        return handle_error("Failed to manage directory", e)
    try:
        lines, err = read_lines_from_directory(prompts_en_dir)
        if err:
            return handle_error("Failed to read fragments", err)
        asyncio.run(async_generate_images(lines))
    except Exception as e:
        return handle_error("Failed to read fragments", e)
    return jsonify({"status": "Image generation started"}), 200

def get_local_images():
    try:
        files = os.listdir(image_dir)
    except Exception as e:
        return jsonify({"error": "Failed to read image directory"}), 500

    image_map = {}
    for file in files:
        if not os.path.isdir(file):
            matches = re.match(r'(\d+)\.png', file)
            if matches:
                key = matches.group(1)
                abs_path = os.path.join("/images", file)
                image_map[key] = abs_path

    return jsonify(image_map), 200

def generate_single_image():
    try:
        req = request.get_json()
        if not req or 'index' not in req or 'content' not in req:
            return jsonify({"error": "parse request body failed"}), 400
        file = os.path.join("/images", str(req['index'])+'.png')
        asyncio.run(async_generate_image_single(req['content'], req['index']))
    except Exception as e:
        return handle_error("Failed to read fragments", e)
    return jsonify({"status": "Image generation started", "url":file}), 200
